Discussion post cleanup (in preparation for post author images support)

marcotuts opened the pull request:
**Description**
A cleanup of the discussion forum styling in preparation for adding in profile images to posts. While the screenshots show images, this detail was removed from the final PR, as it requires follow-up engineering work. 
 
- - -
 
**JIRA Story** ([UX-2586](https://openedx.atlassian.net/browse/UX-2586))
**Confluence / Product Asset** N/A 
**Sandbox URL** (Last Updated: 9/21, 11am) ([pr9666.sandbox.edx.org](http://pr9666.sandbox.edx.org)) 
**Dependencies** N/A
 
- - -
**In Progress Cleanup Changes List** 
A number of forums bugs and visual issues are addressed in this cleanup, in addition to what was necessary to preparation for discussion forum post images. This was done in part to reduce the amount of churn on reviews and other issues from resolving these visual issues in separate PRs. In order to provide more clarity on what the changes were, I've detailed them here below.

SASS Breakdown / General Cleanup
* Moved response and comment styles out of the deprecated _discussion.scss into _response.scss, with some cleanup done to bring these styles in line with our sass style guidelines
* Moved styles for the discussion home grid out of the deprecated _discussion.scss as well. 
* In general, moved rules that weren't correctly using bi-app rtl friendly rules on the discussion forums over to using the correct includes.
* Addition of classes to the underscore templates to move away from element based styles within threads. 

Editing Posts, Responses, Comments
* Removed min-width rule (and other similar rules) causing styling issues on editing of posts. 
* Changed default (pre-focus) height of the edit input area to be the same as the height once focus is applied.

Post Area
* Moved cohort visibility label ("This post is visible to...") section to the post-context area underneath the post content, grouping this with the link to courseware alignment ("Related to:") This cleans up the post header area in preparation for author images.

Response Area
* Reduced the minimum height of responses from 240px to 100px or so, which shortens the visual height of responses that are 1-3 lines long.  

- - -
**PR Author(s) Notes / To-Do** A list of known open tasks that the PR author has.
- [x] test on course with all cases / states for discussion post
- [x] test editability of posts with new styling changes
- [x] remove out commit for images from PR (will be added on in a future TNL story.) 
- [x] rtl testing for changes (and even existing forum styling that was impacted indirectly.) 
- [x] determine next steps for getting test changes onto this branch / PR. 
- [x] fix comment sizing (74%) & response sizing, currently overlaps with actions.
- [x] fix "0 votes" label alignment
- [x] BUG - On create of a post, the left sidebar visually hides other posts excepting the one created. refresh fixes the issue but this seems to not be a problem on production.' 
 
- - -
**Screenshots** 

Post Changes (with static, example post image placeholder)
![image](https://cloud.githubusercontent.com/assets/2023680/9997478/56ed32f6-605a-11e5-9799-93b2bf538d9b.png)


Edit Post Changes (visual fix)
![image](https://cloud.githubusercontent.com/assets/2023680/9751503/c0400594-5673-11e5-9cbe-e2ce5c1750a2.png)


Response Changes
![image](https://cloud.githubusercontent.com/assets/2023680/9997563/c7321ea0-605a-11e5-88a5-44b0d52baaf0.png)

Edit Response Changes (visual fix) 
![image](https://cloud.githubusercontent.com/assets/2023680/9997522/9a3e04fe-605a-11e5-8d69-9822046f7bf7.png)

Comment Changes
![image](https://cloud.githubusercontent.com/assets/2023680/9997530/a3d2dcec-605a-11e5-843f-f1c985da0cf9.png)



- - -
**Reviewers**
- [x] FED: @talbs 
- [ ] Product: @explorerleslie 
- [ ] Documentation: @catong 
 
 - - -
**Authors**
@marcotuts

marcotuts commented on common/static/common/templates/discussion/forum-action-answer.underscore:
```diff
@@ -5,6 +5,6 @@
             <span class="label-checked"><%- gettext("Unmark as Answer") %></span>
         </span>
-        <span class="action-icon"><i class="icon fa fa-ok"></i></span>
+        <span class="action-icon"><i class="icon fa fa-check"></i></span>
```
Context: The first two files are changed to fix a visual regression that caused the "endorse" and "mark as answer" icons to have no icon.

Screenshot:
![image](https://cloud.githubusercontent.com/assets/2023680/9751723/43fe96f0-5676-11e5-8ca5-8339a5eeefb5.png)

marcotuts commented on common/static/common/templates/discussion/response-comment-show.underscore:
```diff
@@ -1,4 +1,7 @@
-<div id="comment_<%- id %>">
+<div class="discussion-comment" id="comment_<%- id %>">
+  <div class="author-image level-comment">
+    <img class="image-frame" src="https://d1kekzok76m982.cloudfront.net/prod/5d0da5229e4c83894a51e04a41e06cd0_120.jpg?v=1441780917" alt="Profile image for post author">
```
Note: Prior to merge, I will remove this img element (if empty, this div is collapsed currently.) I've noted this in the PR checklist as well.

talbs replied:
Thanks for the 411. Makes sense.

talbs replied:
Also, will the ``Profile image for post author`` alt value show the real user's username at runtime?

marcotuts commented on common/static/common/templates/discussion/thread-show.underscore:
```diff
@@ -1,21 +1,10 @@
 <div class="discussion-post">
-    <header>
-        <div class="group-visibility-label">
```
Context: The group-visibility-label div was moved to further down in the post area. Instead of adding this to the top of the post, is has moved into the "post-context" area under the post-body area. This area also conveys when a post "relates to" a specific area of the courseware. Moving the element was done in order to clean up the post-details area, making the visual focus rest on the image / title of the post.

marcotuts commented on common/static/common/templates/discussion/thread-show.underscore:
```diff
@@ -59,16 +48,15 @@
-        );
-        %>
-        <div class="post-context">
+    <div class="post-context">
```
Context: This change is less drastic than github makes it seem. The "post-context" div was moved outside of the conditional logic that exposes the "related to" part of the post-context area, to account for the .group-visibility-label div moving here.

marcotuts commented on common/static/sass/_mixins.scss:
```diff
@@ -276,10 +276,10 @@
   text-align: center;
   border-radius: ($baseline/4);
-  border: 1px solid $blue-l2;
+  //border: 1px solid $blue-l2;
```
Author To-Do: This file's changes should be reverted, as it was unintentionally included from earlier debugging of a sass compilation issue on master.

marcotuts commented on lms/static/sass/_build-lms.scss:
```diff
@@ -63,6 +63,7 @@
 @import "discussion/elements/editor";
 @import "discussion/elements/labels";
 @import "discussion/elements/navigation";
+@import "discussion/views/home";
```
This new partial file removes a section of the deprecated _discussion.scss file (shown in the next file below) to help clarify that these are still actively used styles. The new partial accounts for existing lines 256-416 in _discussion.scss

marcotuts commented on lms/static/sass/discussion/_discussion.scss:
```diff
@@ -423,36 +261,6 @@ body.discussion {
       word-wrap: break-word;
     }
 
-    h1 {
```
Context: These styles (426-455) were also removed from this deprecated partial and into _thread.scss under the "thread - elements - shared styles" section. Instead of using the h1 element it now uses the appropriate (new) classes (.discussion-post, .discussion-response, .discussion-comment)

talbs replied:
Sweet!

marcotuts commented on lms/static/sass/discussion/_discussion.scss:
```diff
@@ -463,123 +271,6 @@ body.discussion {
     margin-bottom: $baseline;
   }
 
-  .responses {
```
Context: These styles (466-582) were also removed from this deprecated partial and moved into _response.scss under a few different sections. I've noted what each section is here, ordered based on how they were defined in _discussion.scss. Their order is slightly adjusted now to aid readability in _responses.scss, and so I've added the table of contents section it applies to in parentheticals. 

First, the layout styling for responses (+wrapper - response list), the next bit is for the banners and labels applied to posts (+response - labels and banners), then the base styles for a single response element (+base - single response element), and finally the add response button (+element - add response area).

marcotuts commented on lms/static/sass/discussion/_discussion.scss:
```diff
@@ -784,13 +475,13 @@ body.discussion {
           overflow-y: auto;
           max-height: 600px;
 
+          //TO-DO: review / remove this from inline styling case 
```
Author To-Do: Review inline styling module to ensure there are no visual or interaction regressions. Noted in PR checklist, will remove this note after that testing as well.

marcotuts commented on lms/static/sass/discussion/_discussion.scss:
```diff
@@ -808,35 +499,13 @@ body.discussion {
-                margin-bottom: 0px !important; // Override courseware CSS
-              }
 
               h4 {
```
Author To-Do: I couldn't find where this h4 is used, which is why this whole section isn't gone yet. I will look again to see where this might be used prior to merging.

marcotuts commented on lms/static/sass/discussion/_discussion.scss:
```diff
@@ -1066,12 +735,3 @@ body.discussion {
   }
 }
 
-// ====================
```
Context: this section was moved to line 94 of _thread.scss, under the area labeled by the following comment: // thread - header content details.

marcotuts commented on lms/static/sass/discussion/_discussion.scss:
```diff
@@ -808,35 +499,13 @@ body.discussion {
 
+            // deprecated styles - remove after review / cleanup
             header {
-              padding-bottom: 0;
```
Context:  These styles were also moved to _thread.scss. They are shared across posts, responses, and comments as well. (+thread - elements - shared styles)

marcotuts commented on lms/static/sass/discussion/_mixins.scss:
```diff
@@ -63,7 +63,7 @@
   padding: ($baseline/2);
   width: 100%;
-  height: 240px;
+  height: 125px;
```
Context: On focus of the response / comment input area, we change the height from 240px to 125px. This change sets the initial height of the input area to the same value it is forced to on focus. This eliminates unnecessary visual height for response and comment authoring, as well as the flicker / change users see when they click into this input area.

marcotuts commented on lms/static/sass/discussion/views/_create-edit-post.scss:
```diff
@@ -9,7 +9,6 @@
   margin: 0;
   border-radius: 3px;
   padding: ($baseline*2);
-  min-width: 760px;
```
Context: This fixes a visual regression / issue where the edit state of responses and comments is visually misaligned.

Screenshot:
![image](https://cloud.githubusercontent.com/assets/2023680/9752013/bbafa330-5679-11e5-8a0e-db3b55eb2d90.png)

marcotuts commented on lms/static/sass/discussion/views/_home.scss:
```diff
@@ -0,0 +1,166 @@
+// discussion - home layout
```
As noted earlier, this whole partial is a move operation from styles in the _discussion.scss partial that was cleaned up for other parts of this PR. There are no changes otherwise within this section.

talbs replied:
Is this partial's job just to style the default landing page for discussions or is this more general for both that view and other views/states users navigate into (discussion detail/thread)?

talbs commented on common/static/common/templates/discussion/thread-response-show.underscore:
```diff
@@ -1,4 +1,7 @@
-<header>
+<header class="wrapper-response-header">
+  <div class="author-image level-response">
+    <img class="image-frame" src="https://d1kekzok76m982.cloudfront.net/prod/5d0da5229e4c83894a51e04a41e06cd0_120.jpg?v=1441780917" alt="Profile image for post author">
```
Will the "Profile image for post author" ``alt`` value show the real user's username at runtime?

marcotuts replied:
Ah yeah, it definitely will need to. I'll rip out the whole img element prior to merge, but will be able to reference this commit id for TNL devs whenever they pick up adding in the img .image-frame element.

talbs replied:
Ahh okay, gotcha.

talbs commented on lms/static/sass/discussion/_discussion.scss:
```diff
@@ -21,9 +21,8 @@ body.discussion {
 
```
I noticed that this Sass partial that is deprecated doesn't have any notes/comments alerting folks to that. Can you add something to the top of this file noting that?

marcotuts replied:
+1 to that. it was previously only on _build-lms.scss. Added to this file.

talbs replied:
Thx!

talbs commented on lms/static/sass/discussion/utilities/_variables.scss:
```diff
@@ -6,3 +10,8 @@ $forum-color-following: $blue;
 $forum-color-marked-answer: $green-d1;
+
+// post images
+$post-image-dimension: ($baseline*3);  // image size  + margin
```
Good variables, but can we hint that these control the size more, perhaps something like:

```scss
$post-image-size: [value]; 
```

or

```scss
$post-image-width: [value];
$post-image-height: $post-image-width;  
```

marcotuts replied:
Sure thing, I can make that change! 

( I originally used "-size", but moved to "-dimension" to echo the learner_profile variables we use for the image-frame there.)

talbs replied:
Oh snap! I didn't know about those from the profile area. Keep em synced and good job to start with. Is there a reason why we're not passing any of those values here?

marcotuts replied:
Ah, so because the variable was in "views/learner-profile" (and not necessarily a base variable / utilitiy) I didn't take that other variable directly, but I could move the learner-profile size out to base/variables perhaps? That may make more sense than what I have here now. 

The benefit of that is we could show all image sizes in the platform in 1 place, and reuse sizes when appropriate. (profile-image, header-image, discussion-post, discussion-response, etc.)

talbs replied:
> The benefit of that is we could show all image sizes in the platform in 1 place, and reuse sizes when appropriate. (profile-image, header-image, discussion-post, discussion-response, etc.)

I like that idea a lot (canned profile image sizes + specific overrides per-app that we control). If its not too much work, great, but if it is or you want to do that in a separate PR aside from this, I'm very much cool with that too.

talbs commented on lms/static/sass/discussion/utilities/_variables.scss:
```diff
@@ -6,3 +10,8 @@ $forum-color-following: $blue;
+
+// post images
+$post-image-dimension: ($baseline*3);  // image size  + margin
+$response-image-dimension: ($baseline*2.5); // image size  + margin
```
See [my note about size-based variable above](https://github.com/edx/edx-platform/pull/9666/files#r39048613).

talbs commented on lms/static/sass/discussion/utilities/_variables.scss:
```diff
@@ -6,3 +10,8 @@ $forum-color-following: $blue;
+// post images
+$post-image-dimension: ($baseline*3);  // image size  + margin
+$response-image-dimension: ($baseline*2.5); // image size  + margin
+$comment-image-dimension: ($baseline*2); // image size  + margin
```
See [my note about size-based variable above](https://github.com/edx/edx-platform/pull/9666/files#r39048613).

talbs commented on lms/static/sass/discussion/views/_home.scss:
```diff
@@ -0,0 +1,166 @@
+// ====================
+
+// home - layout
+.container .discussion-body {
```
If the former is true of [my previous comment](https://github.com/edx/edx-platform/pull/9666/files#r39049227), this selector seems pretty generic and not tied to that "home" view. Is there a body-based class or class added to a container that we can be more specific from?

marcotuts replied:
Ah yeah, I can clean that up. Not everything in this existing file was nested within ".blank-slate" which is the existing "home" class. I can easily nest all other styles in this partial within .blank-slate however, which should make the scope clearer. (naming issue aside of "blank-slate"

talbs replied:
> Ah yeah, I can clean that up. Not everything in this existing file was nested within ".blank-slate" which is the existing "home" class. I can easily nest all other styles in this partial within .blank-slate however, which should make the scope clearer. (naming issue aside of "blank-slate"

Makes sense. I think adding some section-level Sass comments would help to discern what's general forums layout styling across all states/views vs. specific views.

talbs commented on lms/static/sass/discussion/views/_home.scss:
```diff
@@ -0,0 +1,166 @@
+.container .discussion-body {
+
+  .blank-slate {
+      padding: 40px 40px 10px;
```
I know you're just moving, but can these values get the ``$baseline`` treatment?

marcotuts replied:
Absolutely. I can make all the _home.scss partial changes you've suggested in a separate fixup commit, to help make clear the move changes (current) and the cleanup changes (fixup hotness.)

talbs commented on lms/static/sass/discussion/views/_home.scss:
```diff
@@ -0,0 +1,166 @@
+        font-size: 18px;
+        color: $black;
+        margin-bottom: ($baseline/4);
+      }
```
NIT: add a line break.

talbs commented on lms/static/sass/discussion/views/_home.scss:
```diff
@@ -0,0 +1,166 @@
+        font-size: 12px;
+        line-height: 1;
+        margin-bottom: $baseline/2;
+      }
```
NIT: add a line break.

talbs commented on lms/static/sass/discussion/views/_home.scss:
```diff
@@ -0,0 +1,166 @@
+// discussion - home layout
+// ====================
+
```
Any need for a table of contents + sections here to match what you're doing in ``/_response.scss``?

talbs commented on lms/static/sass/discussion/views/_response.scss:
```diff
@@ -1,3 +1,100 @@
+  }
+  
+  .posted-by {
+    font-weight: 700;
```
NIT: can we use preset typography font-weight extends?

talbs replied:
Following up, ^ any chance to use that extend

marcotuts replied:
Thanks for catching that. fixed.

talbs commented on lms/static/sass/discussion/views/_thread.scss:
```diff
@@ -1,23 +1,37 @@
     .post-header-content {
       display: inline-block;
-      width: flex-grid(9,12);
+      // width: calc(flex-grid(9,12) - $post-image-dimension);
```
Is this needed still (since its commented out and you have a ``flex-grid``-based width)?

talbs commented on lms/static/sass/discussion/views/_thread.scss:
```diff
@@ -1,23 +1,37 @@
     .post-header-actions {
-      display: inline-block;
       @include float(right);
+      display: inline-block;
```
Hmm, floating and inline-block+vertical align middle. Usually its one method or the other. Are both needed here?

marcotuts replied:
ah yeah, just reordered properties here blindly without noticing that. cleaned up now. :) good catch.

talbs commented on lms/static/sass/discussion/views/_thread.scss:
```diff
@@ -25,21 +39,13 @@ body.discussion, .discussion-module {
       display: inline-block;
       vertical-align: top;
-      width: flex-grid(11,12);
+      // width: calc((flex-grid(11,12)) - $response-image-dimension);
```
Is this needed still (since its commented out and you have a ``flex-grid``-based width)?

marcotuts replied:
removed these commented out lines. thanks!

talbs commented on lms/static/sass/discussion/views/_thread.scss:
```diff
@@ -25,21 +39,13 @@ body.discussion, .discussion-module {
-    .posted-details .username {
-      font-size: inherit;
-    }
+    min-height: ($baseline*5);
```
is this copied over or a new rule you're setting?

talbs commented on lms/static/sass/discussion/views/_thread.scss:
```diff
@@ -106,24 +72,124 @@ body.discussion, .discussion-module {
     }
 
     .comment-actions-list {
+      @include float(right);
```
Hmm, floating and inline-block+vertical align middle. Usually its one method or the other. Are both needed here?

talbs commented on lms/static/sass/discussion/views/_thread.scss:
```diff
@@ -51,53 +57,13 @@ body.discussion, .discussion-module {
       display: inline-block;
       margin-bottom: ($baseline/2);
-      width: flex-grid(10,12);
+      //width: calc((flex-grid(11,12)) - $comment-image-dimension);
```
Is this needed still (since its commented out and you have a ``flex-grid``-based width)?

marcotuts replied:
removed.

talbs commented on lms/static/sass/discussion/views/_thread.scss:
```diff
@@ -106,24 +72,124 @@ body.discussion, .discussion-module {
+    .author-image {
+      @include margin-right($baseline/2);
+      display: inline-block;
+      background-color: $gray-l4;
```
Is this needed (assuming the image will scale to take up the entirety of the dimensions)?

talbs replied:
^ Following up - do we need this?

marcotuts replied:
Ah, sorry I missed this earlier. removed.

talbs commented on lms/static/sass/discussion/views/_thread.scss:
```diff
@@ -106,24 +72,124 @@ body.discussion, .discussion-module {
+
+      // CASE: post image
+      &.level-post {
+        height: $post-image-dimension;
```
Also, more of a user profile-based question, but can we always count on a user profile being square? Do we ask or give our users a heads up that images will be used in a square format?

talbs commented on lms/static/sass/discussion/views/_thread.scss:
```diff
@@ -106,24 +72,124 @@ body.discussion, .discussion-module {
+
+  .discussion-post, .discussion-response, .discussion-comment {
+    
+    // thread - images
```
Not your job, but it would be great if we could use profile-based Sass variables that are shared here and in other parts of the UI that render user profiles (ideally this is done through a user image component sass partial, but variables would be a good start) to make sure things are synced up

talbs commented on lms/static/sass/discussion/views/_thread.scss:
```diff
@@ -106,24 +72,124 @@ body.discussion, .discussion-module {
+      letter-spacing: 0;
+    }
+
+    // post labels
```
Need this section/comment still?

marcotuts replied:
removed!

talbs commented:
@marcotuts, a few general FED things (alongside some of the inline comments I left):

* I know you're moving pre-existing code to better places, so perhaps its not the right time (for the diff's sake) to also address a lot of the Sass/CSS property re-ordering needed in these rules. If you want to sync that up, great. If not, we can do it later as a follow-up.
* Make sure you remove all of your good personal ``TO-DO`` notes from the files once you address them.

talbs commented:
@marcotuts, a few visual things I noticed:

##### Alignment of post/thread control text
![screen shot 2015-09-09 at 10 27 26 am](https://cloud.githubusercontent.com/assets/163763/9765463/50de2b8a-56e2-11e5-87f3-8a2519f65bac.png)
The labels here seem slightly off vertically (this is on master now too). Can you center them vertically a bit better?

- - -

##### Width on Add Comment textarea
![screen shot 2015-09-09 at 10 28 03 am](https://cloud.githubusercontent.com/assets/163763/9765509/871470ec-56e2-11e5-8d70-31ff7a4bb5a3.png)
When testing out some of your ``min-width`` removals, I noticed that this textarea jets out. Can you check to see how the width is set here and make sure it stays in the UI appropriately?

talbs commented:
@marcotuts, looking good. I left you most of my FED/UX-related feedback. Let me know when:

* you tackle those or have thoughts
* address all of your TO-DOs in the code and in this PRs list

I'll take another spin through then. Shout if I can help or if you have any questions in the meantime.

marcotuts commented:
Regarding your two other notes (with screenshots!) I've removed the min-width on the wmd-panel that was causing that overlap, and fixed the text alignment issue within our actions as well. Thanks for the note / catch on those!

talbs commented:
Thanks for the fixes, @marcotuts. Let me know when I should re-review previous comments and your fixup! commits.

marcotuts commented:
@talbs - I've taken care of all your PR feedback, and done additional testing for the inline discussion module, as well as RTL use cases. I did punt on abstracting out all our profile-image styles into it's own base/element for now, though I expect we'll tackle this once we have TNL time to help with the non html/css side of profile images. 

The sandbox is updating right now, but it should be ready for another review in a bit. 

I've left my changes since review in separate fixup! commits, which should hopefully make it easier to review. 

Thanks!

talbs commented:
@marcotuts, thanks for the damn good work. This is looking good. 

* I added a few small follow-up comments to things that didn't seem to be addressed, but they're NITs at best.
* I also verified that all of the visual notes I made were addressed.
* Thanks much for the ``fixup!`` commits. They were helpful to see what's changed and follow your steps.

- - -

#### A few other notes:

> I did punt on abstracting out all our profile-image styles into it's own base/element for now, though I expect we'll tackle this once we have TNL time to help with the non html/css side of profile images.

That sounds good to me. No need to slow down this stuff for that.

It looks like you have some failing lettuce tests, but I think that may need to be addressed by ENG given you've changed some UI (assuming its not just flaky stuff).

- - -

:+1:

talbs commented:
``fixup! marco forgot some things :)`` - best commit message, I am going to steal this from you for my workflow.

My :+1: still stand. let me know how can help you at all with next in getting this to done with ENG. Nice work!

marcotuts commented:
Thanks @talbs for the review! I've now tagged @explorerleslie for review (and the sandbox is now updating for the purposes of that review.) 

I will follow up offline to find out who the right person is to tag for documentation review.

marcotuts commented:
jenkins run bokchoy
jenkins run lettuce

marcotuts commented:
jenkins run bokchoy

marcotuts commented:
jenkins run bokchoy

marcotuts commented:
@explorerleslie, @catong  - The sandbox is updated and ready for review. Here is a direct link. Thanks! http://pr9666.sandbox.edx.org/courses/course-v1:edX+DemoX+Demo_Course/discussion/forum

NOTE- Images updated to the current sandbox to clarify that rendering profile images is not a part of this PR.

marcotuts commented:
I'll check in tomorrow morning with the team, but I'm inclined to merge this in as the product & documentation impacts are small, and this would let the PR go out with next week's release. Thanks!

talbs commented:
[Congrats!](https://media.giphy.com/media/3oEduNIyuas2lrPaso/giphy.gif)
